sessions/defense-mechanisms/README.md

```diff
@@ -0,0 +1,423 @@
+
+### seccomp
+
+**Seccomp** is a mechanism though which an application may transition into a state where the system calls it performs are restricted. The policy, which may act on a whitelist or blacklist model, is described using [eBPF](https://lwn.net/Articles/593476/).
+
+**Seccomp** filters are instated using the `prctl` syscall (`PR_SET_SECCOMP`). Once it is in effect, the application will be effectively sandboxed and the restrictions will be inherited by child processes.
+
+This may severely limit our exploitation prospects in some cases. In the challenges that we have solved during these sessions, a common goal was spawning a shell and retrieving a certain file (the flag). If the exploited binary used a seccomp filter that disallowed the `execve` syscall (used by the `system` library function), this would have thwarted our exploit.
+
+The [seccomp-tools](https://github.com/david942j/seccomp-tools) suite provides tools for analyzing seccomp filters. The `dump` subcommand may be used to extract the filter from a binary at runtime and display it in a pseudocode format:
+
+```bash
+silvia@imladris:/sss/demo$ seccomp-tools dump ./seccomp_example
+ line  CODE  JT   JF      K
+=================================
+ 0000: 0x20 0x00 0x00 0x00000004  A = arch
+ 0001: 0x15 0x00 0x09 0x40000003  if (A != ARCH_I386) goto 0011
+ 0002: 0x20 0x00 0x00 0x00000000  A = sys_number
+ 0003: 0x15 0x07 0x00 0x000000ad  if (A == rt_sigreturn) goto 0011
+ 0004: 0x15 0x06 0x00 0x00000077  if (A == sigreturn) goto 0011
+ 0005: 0x15 0x05 0x00 0x000000fc  if (A == exit_group) goto 0011
+ 0006: 0x15 0x04 0x00 0x00000001  if (A == exit) goto 0011
+ 0007: 0x15 0x03 0x00 0x00000005  if (A == open) goto 0011
+ 0008: 0x15 0x02 0x00 0x00000003  if (A == read) goto 0011
+ 0009: 0x15 0x01 0x00 0x00000004  if (A == write) goto 0011
+ 0010: 0x06 0x00 0x00 0x00050026  return ERRNO(38)
+ 0011: 0x06 0x00 0x00 0x7fff0000  return ALLOW
+```
+
+In the example above we see a filter operating on the whitelist model: it specifies a subset of syscalls that are allowed: `rt_sigreturn`, `sigreturn`, `exit_group`, `exit`, `open`, `read` and `write`.
+
+To install `seccomp-tools` on the Kali VM, use the the `gem` package manager:
+
+```bash
+$ gem install seccomp-tools
+```
+
+## Challenges
+
+Challenges can be found in the `activities/` directory.
+
+### 01-04. Challenges - rwslotmachine[1-4]
+
+All of the challenges in this section are intended to be solved with **ASLR enabled**. However, you are free to disable it while developing your exploit for debugging purposes. You are provided with the needed shared libraries from the remote system.
+
+The challenges are based on the same "application": the binaries expose very similar functionality with minimal implementation differences. Your job is to identify the defense mechanisms in use for each of them and bypass them in order to read a flag from the remote system.
+
+They are numbered in the suggested solving order.
+
+**Hints**:
```

Comment:
Mark these hints as spoilers.

Reply:
Those are not quite Hints. Maybe I should replace the `Hints` title with something like `Info` or `Tips`.